### DISCUSSION TITLE
Comment envoyé une image ou un fichier via un formulaire

### DISCUSSION THREAD

**Discussion by BILLY9500** [PHPMailer/PHPMailer#3065] [General]
Bonjour, cela fait 1 mois que je cherche le moyen d envoyé une image ou un simple fichier par e-mail, et je n y arrive pas.
j ai chercher p^partout sur internet, j'ai essayer des  centaine de code différent  et aucun ne fonctionne , âpres tout ce temps toutes ses journée et nuit passé je ne sais plus comment faire.
pouvez vous m aider svp

voici mon formulaire

```
          <form action="config-contact_upload.php" method="POST">
            <div class="form-group">
              <label class="form-label" for="exampleInputText">Votre nom</label>
              <input class="form-control" name="name" type="text" placeholder="">
            </div>
            <div class="form-group">
              <label class="form-label" for="exampleInputText">Votre prenom</label>
              <input class="form-control" name="prenom" type="text" placeholder="">
            </div>
            <div class="form-group">
              <label class="form-label" for="exampleInputemail">Votre email</label>
              <input class="form-control" name="email" type="email"
                placeholder="">
            </div>

            <div class="form-group">
              <label class="form-label" for="exampleTextarea1">Votre message</label>
              <textarea class="form-control" name="message" cols="3" rows="5"
                placeholder=""></textarea>
            </div>
	<div class="container">
      <div class="card">
        <div class="card-body">
          <div class="file-upload-card">
            <i class="bi bi-file-earmark-arrow-up text-primary display-2"></i>
            <h5 class="mt-2 mb-4">Telecharger votre image</h5>

            <!-- Form -->
           
              <div class="form-file">
                <input class="form-control d-none" id="my-photo" type="file">
				  <input type="file" id="file" name="myFile" />
                <label class="form-file-label justify-content-center" for="customFile2">
                  <span class="form-file-button btn btn-primary shadow w-100">Selectionnee</span>
                </label>
              </div>
          

            <h6 class="mt-4 mb-0">Fichier supporter</h6>
            <small>.jpg .png .jpeg .gif</small>
          </div>
        </div>
      </div>
    </div>
            <button class="btn btn-primary w-100 d-flex align-items-center justify-content-center" type="submit">
              Envoy&eacute; votre message
              <i class="bi bi-arrow-right fz-16 ms-1"></i>
            </button>
          </form>
```

Voici mon fichier: config-contact_upload.php

```
<?php
$name = $_POST['name'];
$prenom = $_POST['prenom'];
$email = $_POST['email'];
$message = $_POST['message'];
$customFile2 = ($_FILES['customFile2']);
//Load Composer's autoloader
#require 'vendor/autoload.php';
require '/var/www/sites/monsite/public_html/PHPMailer/src/Exception.php';
require '/var/www/sites/monsite/public_html/PHPMailer/src/PHPMailer.php';
require '/var/www/sites/monsite/public_html/PHPMailer/src/SMTP.php';

//Import PHPMailer classes into the global namespace
//These must be at the top of your script, not inside a function
use PHPMailer\PHPMailer\PHPMailer;
use PHPMailer\PHPMailer\SMTP;
use PHPMailer\PHPMailer\Exception;

//Create an instance; passing `true` enables exceptions
$mail = new PHPMailer(true);

try {
//Server settings
$mail->SMTPDebug = SMTP::DEBUG_OFF; //Enable verbose debug output
$mail->isSMTP(); //Send using SMTP
$mail->Host = 'mail-mon_hebergeur.net'; //Set the SMTP server to send through

$mail->SMTPAuth = true; //Enable SMTP authentication
$mail->Username = 'test@monsite.fr'; //SMTP username
$mail->Password = 'xxxxxxxx'; //SMTP password
$mail->SMTPSecure = PHPMailer::ENCRYPTION_SMTPS; //Enable implicit TLS encryption
$mail->Port = 465; //TCP port to connect to; use 587 if you have set `SMTPSecure = PHPMailer::ENCRYPTION_STARTTLS`

//Recipients
$mail->setFrom('contact@monsite.fr', 'test');
$mail->addAddress('monsite@gmail.com', 'Test wd');

//Attachments
#$mail->addAttachment('/var/tmp/file.tar.gz'); //Add attachments
#$mail->addAttachment('/tmp/image.jpg', 'new.jpg'); //Optional name

//Content
$mail->isHTML(true); //Set email format to HTML
$mail->Subject = 'Contact';

$mail->Body = "
    <html>
    <h2><b>".$name." ".$prenom."  ".$email." ".$message." ".$customFile2."         </b></h2>
    </html>"; 



$mail->AltBody = 'This is the body in plain text for non-HTML mail clients';


$mail->send();
  echo ' message envoye';
} catch (Exception $e) {
  echo "Erreur: Votre message n as pas &eacute;tait envoy&eacute;: {$mail->ErrorInfo}";
}

```

**Comment by LECUYER-David** [comment]
Bonjour Billy,

chez moi tout est ok, pour envoyer une image ou un formulaire.
évidemment chacun a son code de fonctionnement.

bref quand tu insères une image dans le body,il faut aussi déclarer un cid, c'est pas lié a un format html=> c'est propre à un envoie de mail par phpmailer.

dans mon cas je récupère tout le contenu d'un message html par la variable $texte_mail.
dedans tu vas avoir:
$texte_mail='<span style:color:red>toto</span><img src="data:image/webp;base64 etc.....

************************
il faut transformer tes images en base64 et les identifiés avec un cid 
PS:moi dans mon $tex_mail mes images sont déja en base 64.

voici mon code 
for ($i=0;$i<100;$i++){
	$image64=trouver_ma_chaine($texte_mail, '<img src="data:image/webp;base64,', '" alt=" "');
	if ($image64!=""){		
	$id = 'img_'.$i;
	$decodedData = base64_decode($image64);	
	$mail->AddStringEmbeddedImage($decodedData,$id,$id,"base64","image/webp" );
	$texte_mail=str_replace('data:image/webp;base64,'.$image64,'cid:'.$id ,$texte_mail);
	}else{	
	break;	
	}
}
for ($i=100;$i<200;$i++){
	$image64=trouver_ma_chaine($texte_mail, '<img src="data:image/svg+xml;base64,', '" alt=" "');
	if ($image64!=""){		
	$id = 'img_'.$i;
	$decodedData = base64_decode($image64);	
	$mail->AddStringEmbeddedImage($decodedData,$id,$id,"base64","image/svg+xml" );
	$texte_mail=str_replace('data:image/svg+xml;base64,'.$image64,'cid:'.$id ,$texte_mail);
	}else{	
	break;	
	}
}
for ($i=200;$i<300;$i++){
	$image64=trouver_ma_chaine($texte_mail, '<img src="data:image/avif;base64,', '" alt=" "');
	if ($image64!=""){		
	$id = 'img_'.$i;
	$decodedData = base64_decode($image64);	
	$mail->AddStringEmbeddedImage($decodedData,$id,$id,"base64","image/avif" );
	$texte_mail=str_replace('data:image/avif;base64,'.$image64,'cid:'.$id ,$texte_mail);
	}else{	
	break;	
	}
}
**************
je vais rechercher dans mon $texte_mail par paquet de 100images si besoin en webp,avif et svg tu peux le faire aussi en jpg et png.
mais moi dans mon $text_mail cela arrives déjà optimisé en ces format et en base64.

plus qu'à finir en faisant attention au alt important quand l'image se charge mal au mail du client et envoyé le tout au body.
n'oublie pas de mettre la function trouver_ma_chaine.

Bon courage, et tiens moi au courant.
David


$texte_mail=str_replace('alt=" "','alt="image en ligne"',$texte_mail);

$mail->Body    =$texte_mail;	


function trouver_ma_chaine($string, $start, $end){
		$string = ' ' . $string;
		$ini = strpos($string, $start);
		if ($ini == 0) return '';
		$ini += strlen($start);
		$len = strpos($string, $end, $ini) - $ini;
		return substr($string, $ini, $len);
	}

**Comment by BILLY9500** [comment]
Bonjour,
merci pour les informations.

Par contre je n'arrive pas a envoyé de mail a l auteur

j'ai un formulaire de contact
lorsque la personne rempli et envoi le formulaire, je reçois bien l e mail.
mais l' auteur du message ne le recois pas

j'aimerai que lorsque l' utilisateur valide le formulaire, qu il reçoit
également cet e-mail avec le sujet et message comme suit
$sujet
$message
merci pour votre aide
voici mon code


<?php
$name = $_POST['name'];
$prenom = $_POST['prenom'];
$email = $_POST['email'];
$sujet = $_POST['sujet'];
$message = $_POST['message'];

//Load Composer's autoloader
#require 'vendor/autoload.php';
require '/var/www/sites/public_html/PHPMailer/src/Exception.php';
require '/var/www/sites/public_html/PHPMailer/src/PHPMailer.php';
require '/var/www/sites/public_html/PHPMailer/src/SMTP.php';

//Import PHPMailer classes into the global namespace
//These must be at the top of your script, not inside a function
use PHPMailer\PHPMailer\PHPMailer;
use PHPMailer\PHPMailer\SMTP;
use PHPMailer\PHPMailer\Exception;

//Create an instance; passing `true` enables exceptions
$mail = new PHPMailer(true);

try {
//Server settings
$mail->SMTPDebug = SMTP::DEBUG_OFF; //Enable verbose debug output
$mail->isSMTP(); //Send using SMTP
$mail->Host = 'mail-a.mon-hebergement.net'; //Set the SMTP server to send
through

$mail->SMTPAuth = true; //Enable SMTP authentication
$mail->Username = ***@***.***'; //SMTP username
$mail->Password = 'xxxxxxxxxxx'; //SMTP password
$mail->SMTPSecure = PHPMailer::ENCRYPTION_SMTPS; //Enable implicit TLS
encryption
$mail->Port = 465; //TCP port to connect to; use 587 if you have set
`SMTPSecure = PHPMailer::ENCRYPTION_STARTTLS`

//Recipients
***@***.***', 'test');
***@***.***', 'wd');

//Attachments
#$mail->addAttachment('/var/tmp/file.tar.gz'); //Add attachments
#$mail->addAttachment('/tmp/image.jpg', 'new.jpg'); //Optional name

//Content
$mail->isHTML(true); //Set email format to HTML
$mail->Subject = 'Contact';

$mail->Body = "
    <html>
<span style=&quot;font-family:arial,helvetica,sans-serif&quot;><strong><br>
Nom:</strong> ".$name."</span><br>
<span
style=&quot;font-family:arial,helvetica,sans-serif;font-size:10pt&quot;>
<strong>Pr&eacute;nom :</strong> ".$prenom."</span><br>
<span
style=&quot;font-family:arial,helvetica,sans-serif;font-size:10pt&quot;>
<strong>E-mail :</strong> ".$email."</span><br>
<span
style=&quot;font-family:arial,helvetica,sans-serif;font-size:10pt&quot;>
<strong>Sujet :</strong> ".$sujet."</span><br>
<span
style=&quot;font-family:arial,helvetica,sans-serif;font-size:10pt&quot;><strong>
Message:</strong> ".$message."</span></p>
</html>";



$mail->AltBody = 'This is the body in plain text for non-HTML mail clients';


$mail->send();
  echo '
M&eacute;ssage envoy&eacute;
</html>';
} catch (Exception $e) {
  echo "Erreur: Votre message n as pas &eacute;tait envoy&eacute;:
{$mail->ErrorInfo}";
}

Le mer. 14 août 2024 à 19:42, LECUYER-David ***@***.***> a
écrit :

**Comment by LECUYER-David** [comment]
Bonjour Billy

Je t'ai préparé tout ce que tu as besoin, mais cela fait en partie de mon projet que j'ai réalisé.

Si cela ne te dérange pas, je préféré en MP.

voici ma boîte mail lecuyerd5@gmail.com, envoie moi un message et je pourrai plus facilement répondre a tes besoins.

Bon courage

cordialement
David LECUYER
